subsys/nrf_security/src/drivers/cracen/cracenpsa/src/ed25519.c

```diff
@@ -0,0 +1,276 @@
+			  size_t message_length, bool is_message)
+{
+	char hashedmessage[SX_ED25519_DGST_SZ];
+	int status;
+
+	if (is_message) {
+		status = hash_input(message, message_length, hashedmessage);
+		if (status != SX_OK)
+			return status;
+
+		return ed25519_sign_internal(priv_key, signature, hashedmessage, SX_ED25519_DGST_SZ,
+					     true);
+	} else {
+		return ed25519_sign_internal(priv_key, signature, message, message_length, true);
+	}
+}
+
+static int ed25519_verify_internal(const uint8_t *pub_key, const char *message,
+				   size_t message_length, const char *signature, bool prehash)
+{
+	int status;
+	char digest[SX_ED25519_DGST_SZ];
+	size_t ed25519_sz = SX_ED25519_SZ;
+	size_t offset = prehash ? 0 : 1;
+	size_t input_count = 4 - offset;
+
+	char const *hash_array[] = {dom2, signature, (const char *)pub_key, message};
+	size_t hash_array_lengths[] = {sizeof(dom2), ed25519_sz, ed25519_sz, message_length};
+
+	status = hash_all_inputs(&hash_array[offset], &hash_array_lengths[offset], input_count,
+				 &sxhashalg_sha2_512, digest);
+	if (status != SX_OK)
+		return status;
+
+	status =
+		sx_ed25519_verify((struct sx_ed25519_dgst *)digest, (struct sx_ed25519_pt *)pub_key,
+				  (const struct sx_ed25519_v *)(signature + SX_ED25519_SZ),
+				  (const struct sx_ed25519_pt *)signature);
+
+	return status;
+}
+
+int cracen_ed25519_verify(const uint8_t *pub_key, const char *message, size_t message_length,
+			  const char *signature)
+{
+	return ed25519_verify_internal(pub_key, message, message_length, signature, false);
+}
+
+int cracen_ed25519ph_verify(const uint8_t *pub_key, const char *message, size_t message_length,
+			    const char *signature, bool is_message)
+{
+	int status;
+	char message_digest[SX_ED25519_DGST_SZ];
+
+	if (is_message) {
+		status = hash_input(message, message_length, message_digest);
+		if (status != SX_OK)
+			return status;
+
+		return ed25519_verify_internal(pub_key, message_digest, SX_ED25519_DGST_SZ,
+					       signature, true);
+	}
+	return ed25519_verify_internal(pub_key, message, message_length, signature, true);
+}
+
+int cracen_ed25519_create_pubkey(const uint8_t *priv_key, uint8_t *pub_key)
+{
+	int status;
+	char digest[SX_ED25519_DGST_SZ];
+	char *pub_key_A = digest + SX_ED25519_SZ;
+
+	status = ed25519_hash_privkey(digest, priv_key);
+	if (status != SX_OK) {
+		return status;
+	}
+	/* The secret scalar s is computed in place from the first half of the
+	 * private key digest.
+	 */
+	decode_scalar_25519(digest);
+
+	/* Clear second half of private key digest: ed25519_ptmult()
+	 * works on an input of SX_ED25519_DGST_SZ bytes.
+	 */
+	safe_memset(pub_key_A, SX_ED25519_SZ, 0, SX_ED25519_SZ);
+
+	/* Perform point multiplication A = [s]B, to obtain the public key A. */
+	status = sx_ed25519_ptmult(
+		(const struct sx_ed25519_dgst *)digest,
+		(struct sx_ed25519_pt *)pub_key_A);
+
+	if (status != SX_OK)
+		return status;
+
+	memcpy(pub_key, pub_key_A, SX_ED25519_SZ);
```

Comment:
Does it make sense to clear the scalar stored in the first part of the digest buffer here?

Maybe the workbuf in the other functions is also worth considering to clear before exiting the function.

Reply:
That does make sense